doc/source/ray-air/examples/pytorch_resnet_batch_prediction.ipynb

```diff
@@ -0,0 +1,398 @@
+     "output_type": "stream",
+     "text": [
+      "Read progress: 100%|██████████| 50/50 [00:09<00:00,  5.42it/s]\n",
+      "Read progress: 100%|██████████| 50/50 [00:00<00:00, 3996.02it/s]\n"
+     ]
+    }
+   ],
+   "source": [
+    "#@title +\n",
+    "if SMOKE_TEST:\n",
+    "    ds = ds.limit(1000)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Notice that the label for images are their corresponding directory names (e.g. n01728920). To find the indices of our model output that correspond to these names, we'll need to download a mapping from the s3 bucket (`imagenet_class_index.json`). The actual class names can also be found from this file."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 19,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import json\n",
+    "import boto3\n",
+    "from botocore import UNSIGNED\n",
+    "from botocore.client import Config\n",
+    "\n",
+    "# Download mapping file from S3\n",
+    "s3 = boto3.client(\"s3\", config=Config(signature_version=UNSIGNED))\n",
+    "s3.download_file(\n",
+    "    \"air-example-data-2\",\n",
+    "    \"imagenette2/imagenet_class_index.json\",\n",
+    "    \"/tmp/imagenet_class_index.json\",\n",
+    ")\n",
+    "\n",
+    "# Build mappings\n",
+    "idx_to_class = json.load(open(\"/tmp/imagenet_class_index.json\", \"r\"))\n",
+    "class_to_idx = {cls_name: int(index) for index, (cls_name, _) in idx_to_class.items()}\n",
+    "idx_to_class_name = {int(index): text for index, (_, text) in idx_to_class.items()}\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Next, let's define a preprocessor to crop and normalize images, as well as convert the class names to indices with the map that we just constructed. We'll use a {class}`BatchMapper <ray.data.preprocessors.BatchMapper>` to implement this custom preprocessing logic. `BatchMapper` takes in a user defined function and applies that as the data transformation to the input dataset, before feeding it into the model."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "import pandas as pd\n",
+    "from torchvision import transforms\n",
+    "from ray.data.preprocessors import BatchMapper\n",
+    "from typing import Dict\n",
+    "\n",
+    "\n",
+    "def preprocess(batch: Dict[str, np.ndarray]) -> Dict[str, np.ndarray]:\n",
+    "    transform = transforms.Compose(\n",
+    "        [\n",
+    "            transforms.ToPILImage(),\n",
+    "            transforms.Resize(256),\n",
+    "            transforms.CenterCrop(224),\n",
+    "            transforms.ToTensor(),\n",
+    "            transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),\n",
+    "        ]\n",
+    "    )\n",
+    "    batch[\"image\"] = np.array([transform(img).numpy() for img in batch[\"image\"]])\n",
```

Comment:
applying the transform 1 image at a time is generally pretty slow...can we batch this transform?

see here for an example: https://github.com/ray-project/ray/blob/50045b610eaa278374980a8602b824e8f5be4929/release/air_tests/air_benchmarks/workloads/gpu_batch_prediction.py#L39-L56

you should just be able to use the same exact logic.

Reply:
honestly, this `to_tensor` function should probably just be included in `TorchVisionPreprocessor`...

Reply:
Btw, this may involve chaining the TorchVisionPreprocessor with BatchMapper.

Internally, datasets will fuse the 2 stages together.

Reply:
Justin and I have tried your logic, but it doesn't work because:

Firstly, `ds.read_images()` will load input images with different shape, therefore the `map_batches()` will not yield a 4-dimensional tensor `[N, H, W, C]`. Instead, it will return an array of np.array: `[[H1, W1, C], [H2, W2, C], ..., [Hn, Wn, C]]`. 
Unfortunately, torchvision.transforms does not support ragged tensor input, so we can't use the lambda function here.

Reply:
If we are resizing everything to 224 anyways, can we specify that in the `size` arg of `read_images` so that we don't have to deal with ragged tensors?

Reply:
Does the `read_images` function crop and rescale the input images to (256, 256)? 

It looks kind of wired because the users have to remove the `transforms.Resize(256)`  operations from their intactly composed pipeline, and resize the input images through our `read_image` api. It's not so intuitive to the new users. But I think this approach could work for our example.

Reply:
it only does the resizing, not cropping or rescaling. We would still need the preprocessor to do those

Reply:
OK! I have changed the code to use the vectorized TorchVisionPreprocessor.